Bottom: 7e0a1279380304de8bf418e71c04ea839128208d
Top:    2350c0375ee7b8473f9d8883b68c3830d17b3b20
Author: Robert Collins <robertc@robertcollins.net>
Date:   2013-03-03 16:12:36 +1300

Refresh of decorate-testcase-result

---

diff --git a/testtools/testcase.py b/testtools/testcase.py
index bac718d..3e78bf1 100644
--- a/testtools/testcase.py
+++ b/testtools/testcase.py
@@ -880,18 +880,21 @@ class DecorateTestCaseResult(object):
         self.before_run = before_run
         self.after_run = after_run
 
+    def _run(self, result, run_method):
+        result = self.callout(result)
+        if self.before_run:
+            self.before_run(result)
+        try:
+            return run_method(result)
+        finally:
+            if self.after_run:
+                self.after_run(result)
+
     def run(self, result=None):
-            result = self.callout(result)
-            if self.before_run:
-                self.before_run(result)
-            try:
-                return self.decorated.run(result)
-            finally:
-                if self.after_run:
-                    self.after_run(result)
+        self._run(result, self.decorated.run)
 
     def __call__(self, result=None):
-            return self.decorated(self.callout(result))
+        self._run(result, self.decorated)
 
     def __getattr__(self, name):
         return getattr(self.decorated, name)
diff --git a/testtools/tests/test_testcase.py b/testtools/tests/test_testcase.py
index fccc482..b14e1bd 100644
--- a/testtools/tests/test_testcase.py
+++ b/testtools/tests/test_testcase.py
@@ -1472,7 +1472,17 @@ class TestDecorateTestCaseResult(TestCase):
             before_run=lambda result: self.log.append('before'),
             after_run=lambda result: self.log.append('after'))
         case.run(None)
-        self.assertEqual([None,
+        case(None)
+        self.assertEqual([
+            None,
+            'before',
+            ('tags', set(), set()),
+            ('startTest', case.decorated),
+            ('addSuccess', case.decorated),
+            ('stopTest', case.decorated),
+            ('tags', set(), set()),
+            'after',
+            None,
             'before',
             ('tags', set(), set()),
             ('startTest', case.decorated),
